### DISCUSSION TITLE
[By Design] cyclesDone property never reaches the total number of cycles

### DISCUSSION THREAD

**Discussion by ghost** [KyryloKuzyk/PrimeTween#116] [General] [answered]
In the file ReusableTween.cs, at function calcTFromElapsedTimeTotal, there is a float precision problem with the conversion from float to int. Currently, the code is:

```
cyclesDone = (int)(_elapsedTimeTotal / cycleDuration);
```
This line uses a direct cast from float to int, which only truncates the result of the division, potentially causing incorrect behavior due to float precision issues. For instance, in the scenario where _elapsedTimeTotal = 9.9f and cycleDuration = 3.3f, the result of the division should be 3, but the truncation results in 2, which is incorrect.

Proposed Solution:
To avoid such float precision issues, it would be better to use Mathf.FloorToInt() to ensure correct rounding down of the result:

```
cyclesDone = Mathf.FloorToInt(_elapsedTimeTotal / cycleDuration);
````

Reproduction Case:
```
_elapsedTimeTotal = 9.9f
cycleDuration = 3.3f
```
In this case, the expected value for cyclesDone should be 3, but with the current implementation, it incorrectly gives 2.
![image](https://github.com/user-attachments/assets/6808033c-2eb8-4c50-b3ce-0dd2d2ddc7ad)

**Comment by KyryloKuzyk** [comment, marked as answer]
Hey, thank you for creating this topic!

Casting float to int has the same behavior as Mathf.FloorToInt() for positive numbers and the current implementation is the intended one. 

You can only observe the `cyclesDone == settings.cycles` if you pause the animation and set the `tween.progressTotal = 1f` manually. But under normal circumstances animation completes immediately when it reaches the last cycleDone, so it would not be possible to observe that in the Inspector.

You can think of the cyclesDone the following way: it equals to `0` on the first cycle (no cycles completed yet) and equals to `cycles - 1` on the last cycle. And the exact moment the last cycle completes, the animation is recycled immediately.

**Comment by KyryloKuzyk** [comment]
I reopened the discussion so it can be found easier.
